crates/provider/src/chain/stream.rs

```diff
@@ -0,0 +1,159 @@
+use alloy_network::{Ethereum, Network};
+use alloy_primitives::{BlockNumber, U64};
+use alloy_rpc_client::{BatchRequest, PollerBuilder, RpcClientInner, Waiter, WeakClient};
+use alloy_rpc_types_eth::Block;
+use alloy_transport::Transport;
+use async_stream::stream;
+use futures::{stream::FuturesOrdered, Stream, StreamExt};
+use std::{marker::PhantomData, sync::Arc};
+use tokio_stream::wrappers::BroadcastStream;
+
+use super::ChainStreamError;
+
+/// Stream state.
+#[derive(Debug)]
+pub(super) struct PollerStream<T, N = Ethereum> {
+    client: WeakClient<T>,
+    /// Next block that should be yielded.
+    next_yield: BlockNumber,
+    /// Last known block number.
+    chain_tip: BlockNumber,
+    /// Stream of block numbers, updated with respect to polling interval.
+    block_numbers_stream: BroadcastStream<U64>,
+    _phantom: PhantomData<N>,
+}
+
+impl<T: Transport + Clone, N: Network> PollerStream<T, N> {
+    pub(super) fn stream(
+        client: WeakClient<T>,
+        next_yield: Option<u64>,
+    ) -> impl Stream<Item = Block> + 'static {
+        stream! {
+            let mut self_ = match Self::new(client, next_yield).await {
+                Ok(poller) => poller,
+                Err(err) => {
+                    warn!(%err, "Unable to start the stream");
+                    return;
+                }
+            };
+            loop {
+                let blocks = match self_.new_blocks().await {
+                    Ok(blocks) => blocks,
+                    Err(err) => {
+                        warn!(%err, "irrecoverable error, stopping the stream");
+                        break;
+                    }
+                };
+                for block in blocks {
+                    yield block;
+                    self_.next_yield += 1;
```

Comment:
hmm, this doesn't check block numbers?
but we apparently also didn't check this in the original impl

Reply:
Yup. I think we can say that the invariant here is that `PollerStream::new_blocks` is responsible for making sure that the state of the object is unambiguous, and whatever it produces is a sequential list of blocks that _must_ be yielded.
Otherwise, handling situations like reorgs (which are not implemented yet -- will be done later) will be spread among many different places and it will be hard to manage.
I've added a comment for that.